src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WorkerThread.cs

```diff
@@ -0,0 +1,164 @@
+                        if (oldCounts == counts)
+                        {
+                            HillClimbing.ThreadPoolHillClimber.ForceChange(newCounts.numThreadsGoal, HillClimbing.StateOrTransition.ThreadTimedOut);
+                            // TODO: Event:  Worker Thread stop event
+                            return;
+                        }
+                        counts = oldCounts;
+                    } 
+                }
+            }
+
+            internal static void MaybeAddWorkingWorker()
+            {
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_aligned.counts);
+                ThreadCounts newCounts;
+                while (true)
+                {
+                    newCounts = counts;
+                    newCounts.numProcessingWork = Math.Max(counts.numProcessingWork, Math.Min((short)(counts.numProcessingWork + 1), counts.numThreadsGoal));
+                    newCounts.numExistingThreads = Math.Max(counts.numExistingThreads, newCounts.numProcessingWork);
+                    
+                    if(newCounts == counts)
+                    {
+                        return;
+                    }
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_aligned.counts, newCounts, counts);
+
+                    if(oldCounts == counts)
+                    {
+                        break;
+                    }
+
+                    counts = oldCounts;
+                }
+
+                int toCreate = newCounts.numExistingThreads - counts.numExistingThreads;
+                int toRelease = newCounts.numProcessingWork - counts.numProcessingWork;
+
+                if(toRelease > 0)
+                {
+                    s_semaphore.Release(toRelease);
+                }
+
+                for (int i = 0; i < toCreate; i++)
+                {
+                    CreateWorkerThread();
+                }
+            }
+
+            internal static bool ShouldStopProcessingWorkNow()
+            {
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_aligned.counts);
+                while (true)
+                {
+                    if (counts.numExistingThreads <= counts.numProcessingWork)
+                    {
+                        return true;
+                    }
+
+                    ThreadCounts newCounts = counts;
+                    newCounts.numProcessingWork--;
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_aligned.counts, newCounts, counts);
+
+                    if (oldCounts == counts)
+                    {
+                        return false;
+                    }
+                    counts = oldCounts;
+                }
+            }
+
+            private static void CreateWorkerThread()
+            {
+                RuntimeThread workerThread = RuntimeThread.Create(WorkerThreadStart);
```

Comment:
It seems users of the public API would have the same issue if they try to create a bunch of threads. From the current code, it looks like the only reason it waits is to check for failure cases in the thread entry point (StartThread), which can only happen for RoInitialize when WinRT is enabled (the other stuff in the try block are thread-local stores, not sure if they can OOM). If RoInitialize is the only point of failure, would it be possible to not wait when WinRT is not enabled? CC @AntonLapounov